File: python/kickstarter/data_generator_app.py
from pyflink.common import Types, Row
from pyflink.datastream import StreamExecutionEnvironment
from pyflink.table.table_environment import StreamTableEnvironment
from pyflink.table import TableSchema, DataTypes, TableEnvironment, EnvironmentSettings, ExplainDetail
from pyflink.table.catalog import ObjectPath, CatalogBaseTable
import logging
import os
import sys
import time

from common_functions import get_app_options
from data_generator import DataGenerator


__copyright__  = "Copyright (c) 2024 Jeffrey Jonathan Jennings"
__credits__    = ["Jeffrey Jonathan Jennings"]
__license__    = "MIT"
__maintainer__ = "Jeffrey Jonathan Jennings"
__email__      = "j3@thej3.com"
__status__     = "dev"


# Set up the logger
logger = logging.getLogger('DataGeneratorApp')


class DataGeneratorApp:
    """
    This class generates synthetic flight data for two fictional airlines, Sunset Air
    and SkyOne Airlines. The synthetic flight data is stored in separate Apache Iceberg
    tables for Sunset Air and SkyOne Airlines.
    """
    
    @staticmethod
    def main(args):
        # Create a blank Flink execution environment (a.k.a. the Flink job graph -- the DAG)
        env = StreamExecutionEnvironment.get_execution_environment()

        env.set_parallelism(1)

        # Add the custom source
        data_stream_producer_properties = env.add_source(KafkaClientPropertiesLookup(flag=None, options=get_app_options(False, args)))

        producer_properties = {}
        try:
            for type_value in data_stream_producer_properties.execute_and_collect():
                producer_properties.update(type_value)
        except Exception as e:
            print(f"The Flink App stopped during the reading of the custom data source stream because of the following: {e}")
            exit(1)
        
        # Set up the table environment to work with the Table and SQL API in Flink
        table_env = StreamTableEnvironment.create(env, EnvironmentSettings.in_streaming_mode())

        # Define the placheolder values for the preceeding Flink SQL statements
        catalog_name = "apache_kickstarter"
        database_name = "airlines"
        s3_bucket = os.environ['AWS_S3_BUCKET']
        aws_region = os.environ['AWS_REGION']
        skyone_airlines_table = "skyone_airlines"
        sunset_air_table = "sunset_airlines"

        # Define the CREATE CATALOG Flink SQL statement to register the Iceberg catalog
        # using the HadoopCatalog to store metadata in AWS S3, a Hadoop-compatible 
        # filesystem.  Then execute the Flink SQL statement to register the Iceberg catalog
        table_env.execute_sql(f"""
            CREATE CATALOG {catalog_name} WITH (
                'type'='iceberg',
                'catalog-type'='hadoop',
                'catalog-impl'='org.apache.iceberg.flink.FlinkCatalog',
                'warehouse'='{s3_bucket}',
                'io-impl' = 'org.apache.iceberg.aws.s3.S3FileIO',
                'aws.region' = '{aws_region}',
                's3.endpoint' = 's3.{aws_region}.amazonaws.com'
                );
        """)
        table_env.execute_sql(f"USE CATALOG {catalog_name};")

        # Access the Iceberg catalog to create the airlines database and the Iceberg tables
        catalog = table_env.get_catalog(catalog_name).get()

        # Check if the database exists.  If it does not exist, create the database
        try:
            if not catalog.database_exists(database_name):
                table_env.execute_sql(f"CREATE DATABASE IF NOT EXISTS {database_name};")
                table_env.execute_sql(f"USE {database_name};")
        except Exception as e:
            print(f"A critical error occurred to during the processing of the database because {e}")
            sys.exit(1)
        
        # An ObjectPath in Apache Flink is a class that represents the fully qualified path to a
        # catalog object, such as a table, view, or function.  It uniquely identifies an object
        # within a catalog by encapsulating both the database name and the object name.  For 
        # instance, this case we using it to get the fully qualified path of the `skyone_airlines`
        # table
        skyone_airlines_table_path = ObjectPath(database_name, skyone_airlines_table)

        # Check if the table exists.  If it does not exist, create the table
        try:
            if not catalog.table_exists(skyone_airlines_table_path):
                # Define the table schema
                schema = TableSchema.builder() \
                    .field("email_address", DataTypes.STRING()) \
                    .field("flight_departure_time", DataTypes.TIMESTAMP(3)) \
                    .field("iata_departure_code", DataTypes.STRING()) \
                    .field("flight_arrival_time", DataTypes.TIMESTAMP(3)) \
                    .field("iata_arrival_code", DataTypes.STRING()) \
                    .field("flight_number", DataTypes.STRING()) \
                    .field("confirmation", DataTypes.STRING()) \
                    .field("ticket_price", DataTypes.STRING()) \
                    .field("aircraft", DataTypes.STRING()) \
                    .field("booking_agency_email", DataTypes.STRING()) \
                    .build()
                
                # Define Apache Iceberg-specific table properties
                properties = {
                    'connector': 'iceberg',
                    'catalog-type': 'hadoop',  # Type of Iceberg catalog (used for working with AWS S3)
                    'warehouse': f"'{s3_bucket}'",  # Warehouse directory where Iceberg stores data and metadata
                    'write.format.default': 'parquet',  # File format for Iceberg writes
                    'write.target-file-size-bytes': '134217728',  # Target size for files written by Iceberg (128 MB by default)
                    'partitioning': 'iata_arrival_code'  # Optional: Partitioning columns for Iceberg table
                }

                # Create a CatalogBaseTable instance, which is an instantiated object that represents the
                # metadata of a table within a catalog.  It encapsulates all the necessary information
                # about a table's schema, properties, and characteristics, allowing Flink to interact
                # with various data sources and sinks in a unified and consistent manner
                catalog_table = CatalogBaseTable.create_table(schema=schema, properties=properties, comment="The SkyOne Airlines table")

                # Create the table in the catalog
                catalog.create_table(skyone_airlines_table_path, catalog_table, ignore_if_exists=True)
        except Exception as e:
            print(f"A critical error occurred to during the processing of the table because {e}")
            sys.exit(1)

        # An ObjectPath in Apache Flink is a class that represents the fully qualified path to a
        # catalog object, such as a table, view, or function.  It uniquely identifies an object
        # within a catalog by encapsulating both the database name and the object name.  For 
        # instance, this case we using it to get the fully qualified path of the `sunset_air`
        # table
        sunset_air_table_path = ObjectPath(database_name, sunset_air_table)

        # Check if the table exists.  If it does not exist, create the table
        try:
            if not catalog.table_exists(sunset_air_table_path):
                # Create a CatalogBaseTable instance, which is an instantiated object that represents the
                # metadata of a table within a catalog.  It encapsulates all the necessary information
                # about a table's schema, properties, and characteristics, allowing Flink to interact
                # with various data sources and sinks in a unified and consistent manner
                catalog_table = CatalogBaseTable.create_table(schema=schema, properties=properties, comment="The Sunset Air table")

                # Create the table in the catalog
                catalog.create_table(sunset_air_table_path, catalog_table, ignore_if_exists=True)
        except Exception as e:
            print(f"A critical error occurred to during the processing of the table because {e}")
            sys.exit(1)

        # Insert the row of synthetic flight data into the SkyOne Airlines table
        data_gen = DataGenerator()
        row = data_gen.generate_skyone_airlines_flight_data()
        table_env.from_elements([row.to_row()], schema=schema) \
            .execute_insert(skyone_airlines_table_path)
        
        # Creates a rate limiter that allows emitting one event per second.  This is useful for
        # controlling the rate of data generation in real-world scenarios, preventing overloading
        # downstream operators or external systems, and controlling the flow of data to test system
        # behavior under specific conditions
        time.sleep(1)

        # Insert the row of synthetic flight data into the Sunset Air table
        data_gen = DataGenerator()
        row = data_gen.generate_sunset_air_flight_data()
        table_env.from_elements([row.to_row()], schema=schema) \
            .execute_insert(sunset_air_table_path)
        
        # Creates a rate limiter that allows emitting one event per second.  This is useful for
        # controlling the rate of data generation in real-world scenarios, preventing overloading
        # downstream operators or external systems, and controlling the flow of data to test system
        # behavior under specific conditions
        time.sleep(1)

        # Execute the Flink job graph
        try:
            env.execute("DataGeneratorApp")
        except Exception as e:
            logger.error("The App stopped early due to the following: %s", e)


if __name__ == "__main__":
    DataGeneratorApp.main(sys.argv)

File: java/app/src/main/java/kickstarter/DataGeneratorApp.java
/**
 * Copyright (c) 2024 Jeffrey Jonathan Jennings
 * 
 * @author Jeffrey Jonathan Jennings (J3)
 * 
 * 
 */
package kickstarter;

import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.common.typeinfo.Types;
import org.apache.flink.api.connector.source.util.ratelimit.RateLimiterStrategy;
import org.apache.flink.connector.base.DeliveryGuarantee;
import org.apache.flink.connector.datagen.source.DataGeneratorSource;
import org.apache.flink.connector.kafka.sink.*;
import org.apache.flink.formats.json.JsonSerializationSchema;
import org.apache.flink.streaming.api.datastream.*;
import org.apache.flink.streaming.api.environment.*;
import org.apache.flink.table.api.*;
import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
import org.apache.flink.table.data.*;
import org.apache.flink.table.types.logical.*;
import org.apache.flink.types.*;
import org.apache.flink.table.catalog.*;
import org.apache.iceberg.catalog.*;
import org.apache.iceberg.flink.*;
import org.apache.iceberg.flink.sink.FlinkSink;
import org.apache.hadoop.conf.Configuration;
import java.util.*;
import org.slf4j.*;

import kickstarter.model.*;


/**
 * This class creates fake flight data for fictional airlines <b>Sunset Air</b> and 
 * <b>Sky One</b> Airlines," and sends it to the Kafka topics `airline.sunset` and `airline.skyone`,
 * respectively.
 */
public class DataGeneratorApp {
    private static final Logger logger = LoggerFactory.getLogger(DataGeneratorApp.class);


	/**
	 * The main method in a Flink application serves as the entry point of the program, where
	 * the Flink DAG is defined.  That is, the execution environment, the creation of the data
	 * streams or datasets, apply transformations, and trigger the execution of the application (by
	 * sending it to the Flink JobManager).
	 * 
	 * @param args list of strings passed to the main method from the command line.
	 * @throws Exception - The exceptions are forwarded, and are caught by the runtime.  
     * When the runtime catches an exception, it aborts the task and lets the fail-over logic
	 * decide whether to retry the task execution.
	 */
	public static void main(String[] args) throws Exception {
        /*
         * Retrieve the value(s) from the command line argument(s)
         */
        String serviceAccountUser = Common.getAppArgumentValue(args, Common.ARG_SERVICE_ACCOUNT_USER);
        String awsRegion = Common.getAppArgumentValue(args, Common.ARG_AWS_REGION);

		// --- Create a blank Flink execution environment (a.k.a. the Flink job graph -- the DAG)
		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        // --- Enable checkpointing every 5000 milliseconds (5 seconds)
        env.enableCheckpointing(5000);

        /*
         * Set timeout to 60 seconds
         * The maximum amount of time a checkpoint attempt can take before being discarded.
         */
        env.getCheckpointConfig().setCheckpointTimeout(60000);

        /*
         * Set the maximum number of concurrent checkpoints to 1 (i.e., only one checkpoint
         * is created at a time)
         */
        env.getCheckpointConfig().setMaxConcurrentCheckpoints(1);

        // --- Create a TableEnvironment
        EnvironmentSettings settings = 
            EnvironmentSettings.newInstance()
                               .inStreamingMode()
                               .build();
        StreamTableEnvironment tblEnv = StreamTableEnvironment.create(env, settings);

		/*
		 * --- Kafka Producer Config
		 * Retrieve the properties from AWS Secrets Manager and AWS Systems Manager Parameter Store.
		 * Then ingest properties into the Flink app
		 */
        DataStream<Properties> dataStreamProducerProperties = 
			env.fromData(new Properties())
			   .map(new KafkaClientPropertiesLookup(false, serviceAccountUser))
			   .name("kafka_producer_properties");
		Properties producerProperties = new Properties();

		/*
		 * Execute the data stream and collect the properties.
		 * 
		 * Note, the try-with-resources block ensures that the close() method of the CloseableIterator is
		 * called automatically at the end, even if an exception occurs during iteration.
		 */
		try {
			dataStreamProducerProperties
				.executeAndCollect()
                .forEachRemaining(typeValue -> {
                    producerProperties.putAll(typeValue);
                });
		} catch (final Exception e) {
            System.out.println("The Flink App stopped during the reading of the custom data source stream because of the following: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
		}

        /*
         * Create a data generator source
         */
        DataGeneratorSource<AirlineData> skyOneSource =
            new DataGeneratorSource<>(
                index -> DataGenerator.generateAirlineFlightData("SKY1"),
                Long.MAX_VALUE,
                RateLimiterStrategy.perSecond(1),
                Types.POJO(AirlineData.class)
            );

        /*
         * Sets up a Flink POJO source to consume data
         */
        DataStream<AirlineData> skyOneStream = 
            env.fromSource(skyOneSource, WatermarkStrategy.noWatermarks(), "skyone_source");

        /*
         * Sets up a Flink Kafka sink to produce data to the Kafka topic `airline.skyone` with the
         * specified serializer
         */
        KafkaRecordSerializationSchema<AirlineData> skyOneSerializer = 
            KafkaRecordSerializationSchema.<AirlineData>builder()
                .setTopic("airline.skyone")
                .setValueSerializationSchema(new JsonSerializationSchema<>(Common::getMapper))
                .build();

        /*
         * Takes the results of the Kafka sink and attaches the unbounded data stream to the Flink
         * environment (a.k.a. the Flink job graph -- the DAG)
         */
        KafkaSink<AirlineData> skyOneSink = 
            KafkaSink.<AirlineData>builder()
                .setKafkaProducerConfig(producerProperties)
                .setRecordSerializer(skyOneSerializer)
                .setDeliveryGuarantee(DeliveryGuarantee.AT_LEAST_ONCE)
                .build();

        /*
         * Adds the given Sink to the DAG. Note only streams with sinks added will be executed
         * once the StreamExecutionEnvironment.execute() method is called
         */
        skyOneStream.sinkTo(skyOneSink).name("skyone_sink");

        /*
         * Sets up a Flink POJO source to consume data
         */
        DataGeneratorSource<AirlineData> sunsetSource =
            new DataGeneratorSource<>(
                index -> DataGenerator.generateAirlineFlightData("SUN"),
                Long.MAX_VALUE,
                RateLimiterStrategy.perSecond(1),
                Types.POJO(AirlineData.class)
            );

        DataStream<AirlineData> sunsetStream = 
            env.fromSource(sunsetSource, WatermarkStrategy.noWatermarks(), "sunset_source");

        /*
         * Sets up a Flink Kafka sink to produce data to the Kafka topic `airline.sunset` with the
         * specified serializer
         */
        KafkaRecordSerializationSchema<AirlineData> sunsetSerializer = 
            KafkaRecordSerializationSchema.<AirlineData>builder()
                .setTopic("airline.sunset")
                .setValueSerializationSchema(new JsonSerializationSchema<>(Common::getMapper))
                .build();

        /*
         * Takes the results of the Kafka sink and attaches the unbounded data stream to the Flink
         * environment (a.k.a. the Flink job graph -- the DAG)
         */
        KafkaSink<AirlineData> sunsetSink = 
            KafkaSink.<AirlineData>builder()
                .setKafkaProducerConfig(producerProperties)
                .setRecordSerializer(sunsetSerializer)
                .setDeliveryGuarantee(DeliveryGuarantee.AT_LEAST_ONCE)
                .build();

        /*
         * Adds the given Sink to the DAG. Note only streams with sinks added will be executed
         * once the StreamExecutionEnvironment.execute() method is called
         */
        sunsetStream.sinkTo(sunsetSink).name("sunset_sink");

        String catalogName = "apache_kickstarter";
        String bucketName = serviceAccountUser.replace("_", "-");  // --- To follow S3 bucket naming convention, replace underscores with hyphens if exist
        String warehousePath = "s3://" + bucketName + "/warehouse";
        String catalogImpl = "org.apache.iceberg.aws.glue.GlueCatalog";
        String ioImpl = "org.apache.iceberg.aws.s3.S3FileIO";
        String databaseName = "airlines";

        // --- Configure the AWS Glue Catalog Properties
        Map<String, String> catalogProperties = new HashMap<>();
        catalogProperties.put("type", "iceberg");
        catalogProperties.put("warehouse", warehousePath);
        catalogProperties.put("catalog-impl", catalogImpl);
        catalogProperties.put("io-impl", ioImpl);
        catalogProperties.put("glue.skip-archive", "true");
        catalogProperties.put("glue.region", awsRegion);
        
        // --- Use the CatalogLoader since an external metastore is used (AWS Glue Catalog)
        CatalogLoader catalogLoader = CatalogLoader.custom(catalogName, catalogProperties,  new Configuration(false), catalogImpl);

        // --- Describes and configures the catalog for the Table API and SQL
        CatalogDescriptor catalogDescriptor = CatalogDescriptor.of(catalogName, org.apache.flink.configuration.Configuration.fromMap(catalogProperties));

        // --- Create the catalog, use it, and get the instantiated catalog
        tblEnv.createCatalog(catalogName, catalogDescriptor);
        tblEnv.useCatalog(catalogName);
        org.apache.flink.table.catalog.Catalog catalog = tblEnv.getCatalog("apache_kickstarter").orElseThrow(() -> new RuntimeException("Catalog not found"));

        // --- Print the current catalog name
        System.out.println("Current catalog: " + tblEnv.getCurrentCatalog());

        // --- Check if the database exists.  If not, create it
        try {
            if (!catalog.databaseExists(databaseName)) {
                catalog.createDatabase(databaseName, new CatalogDatabaseImpl(new HashMap<>(), "The Airlines flight data database."), false);
            }
            tblEnv.useDatabase(databaseName);
        } catch (Exception e) {
            System.out.println("A critical error occurred during the processing of the database because " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }

        // --- Print the current database name
        System.out.println("Current database: " + tblEnv.getCurrentDatabase());

        // --- Define the RowType for the RowData
        RowType rowType = RowType.of(
            new LogicalType[] {
                DataTypes.STRING().getLogicalType(),
                DataTypes.STRING().getLogicalType(),
                DataTypes.STRING().getLogicalType(),
                DataTypes.STRING().getLogicalType(),
                DataTypes.STRING().getLogicalType(),
                DataTypes.BIGINT().getLogicalType(),
                DataTypes.STRING().getLogicalType(),
                DataTypes.STRING().getLogicalType(),
                DataTypes.DECIMAL(10, 2).getLogicalType(),
                DataTypes.STRING().getLogicalType(),
                DataTypes.STRING().getLogicalType()
            },
            new String[] {
                "email_address",
                "departure_time",
                "departure_airport_code",
                "arrival_time",
                "arrival_airport_code",
                "flight_duration",
                "flight_number",
                "confirmation_code",
                "ticket_price",
                "aircraft",
                "booking_agency_email"
            }
        );
        
        SinkToIcebergTable(tblEnv, catalog, catalogLoader, databaseName, rowType.getFieldCount(), "skyone_airline", skyOneStream);
        SinkToIcebergTable(tblEnv, catalog, catalogLoader, databaseName, rowType.getFieldCount(), "sunset_airline", sunsetStream);

        try {
            // --- Execute the Flink job graph (DAG)
            env.execute("DataGeneratorApp");
        } catch (Exception e) {
            logger.error("The App stopped early due to the following: {}", e.getMessage());
        }
	}

    /**
     * This method is used to sink the data from the input data stream into the iceberg table.
     * 
     * @param tblEnv The StreamTableEnvironment.
     * @param catalog The Catalog. 
     * @param catalogLoader The CatalogLoader.
     * @param databaseName  The name of the database.
     * @param fieldCount The number of fields in the input data stream.
     * @param tableName The name of the table. 
     * @param airlineDataStream The input data stream.
     */
    private static void SinkToIcebergTable(final StreamTableEnvironment tblEnv, final org.apache.flink.table.catalog.Catalog catalog, final CatalogLoader catalogLoader, final String databaseName, final int fieldCount, final String tableName, DataStream<AirlineData> airlineDataStream) {
        // --- Convert DataStream<AirlineData> to DataStream<RowData>
        DataStream<RowData> skyOneRowData = airlineDataStream.map(new MapFunction<AirlineData, RowData>() {
            @Override
            public RowData map(AirlineData airlineData) throws Exception {
                GenericRowData rowData = new GenericRowData(RowKind.INSERT, fieldCount);
                rowData.setField(0, StringData.fromString(airlineData.getEmailAddress()));
                rowData.setField(1, StringData.fromString(airlineData.getDepartureTime()));
                rowData.setField(2, StringData.fromString(airlineData.getDepartureAirportCode()));
                rowData.setField(3, StringData.fromString(airlineData.getArrivalTime()));
                rowData.setField(4, StringData.fromString(airlineData.getArrivalAirportCode()));
                rowData.setField(5, airlineData.getFlightDuration());
                rowData.setField(6, StringData.fromString(airlineData.getFlightNumber()));
                rowData.setField(7, StringData.fromString(airlineData.getConfirmationCode()));
                rowData.setField(8, DecimalData.fromBigDecimal(airlineData.getTicketPrice(), 10, 2));
                rowData.setField(9, StringData.fromString(airlineData.getAircraft()));
                rowData.setField(10, StringData.fromString(airlineData.getBookingAgencyEmail()));
                return rowData;
            }
        });
        
        TableIdentifier tableIdentifier = TableIdentifier.of(databaseName, tableName);

        // Create the table if it does not exist
        if (!catalog.tableExists(ObjectPath.fromString(databaseName + "." + tableName))) {
            tblEnv.executeSql(
                        "CREATE TABLE " + databaseName + "." + tableName + " ("
                            + "email_address STRING, "
                            + "departure_time STRING, "
                            + "departure_airport_code STRING, "
                            + "arrival_time STRING, "
                            + "arrival_airport_code STRING, "
                            + "flight_duration BIGINT,"
                            + "flight_number STRING, "
                            + "confirmation_code STRING, "
                            + "ticket_price DECIMAL(10,2), "
                            + "aircraft STRING, "
                            + "booking_agency_email STRING) "
                            + "WITH ("
                                + "'write.format.default' = 'parquet',"
                                + "'write.target-file-size-bytes' = '134217728',"
                                + "'partitioning' = 'arrival_airport_code',"
                                + "'format-version' = '2');"
                    );
        }

        // ---
        TableLoader tableLoaderSkyOne = TableLoader.fromCatalog(catalogLoader, tableIdentifier);

        FlinkSink
            .forRowData(skyOneRowData)
            .tableLoader(tableLoaderSkyOne)
            .upsert(true)
            .equalityFieldColumns(Arrays.asList("email_address", "departure_airport_code", "arrival_airport_code"))
            .append();
    }
}